### ISSUE TITLE
Changing the root page when navigating crashes the application on Android

### ISSUE THREAD

**Issue by vhugogarcia** [dotnet/maui#10824]
### Description

Hello friends,

First, I would like to THANK YOU all for #dotnetmaui, being honest, it has been a delight developing apps using .NET MAUI. It is fast, reliable and overcool! Developing UI using Borders is probably my favorite feature ever in the world! It is so much easy now to style form controls or any other element in the UI with Borders, it is just amazing.

Let me start detailing the issue, I'm facing.

**Note:** I'm using Prism.Maui on the application, I know it is still on beta, however, I believe this issue is related to MAUI and not Prism itself. 

The application has defined a flow for non-logged users and logged users. Any flow starts with the OnBoarding page, this page provides just a welcome message to the users and it is in charge of redirect the logged users to the Dashboard page. However, if the user is at the OnBoarding screen, and for example _(just for the demo)_ I click on Sign In button and want to redirect the user to the Dashboard screen and for instance remove from the navigation stack the root page and now Dashboard become the root page, so we prevent logged the users to navigate back while the app is open.

So, when navigating to a page we change the root and on Android it appears to do a jump into another app somehow and for instance it generates an issue. I'm sharing a video to demo it below:

https://user-images.githubusercontent.com/1047398/197197951-df6299de-d96e-4e7c-a109-12cd80791703.mp4

  
<img width="1036" alt="image" src="https://user-images.githubusercontent.com/1047398/197198810-13689fa5-c6e3-4217-a978-ed4e1c635ca8.png">

### Steps to Reproduce

1. Create a new Maui Application
2. Add Prism.Maui to the project and setup the initialization
3. Create a 2 Views and 2 ViewModels
4. Register them on Prism.Maui
5. Navigate from one page to another page (Not modal)

### Link to public reproduction project repository

https://github.com/vhugogarcia/acceso/tree/feature/base-architecture

### Version with bug

7.0 Release Candidate 2

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android (any version)

### Did you find any workaround?

I could not find a workaround at this moment.

### Relevant log output

```shell

The code I'm using to navigate to another page is:
 _navigationService.NavigateAsync(uri)
            .OnNavigationError(ex => Debug.WriteLine(ex));

I believe this is the source code from Prism.Maui: https://github.com/PrismLibrary/Prism.Maui/blob/master/src/Prism.Maui/Navigation/PageNavigationService.cs

Error Message: Either set MainPage or override CreateWindow.
Error Source: Microsoft.Maui.Controls
Error StackTrace:
   at Microsoft.Maui.Controls.Application.CreateWindow(IActivationState activationState) in D:\a\_work\1\s\src\Controls\src\Core\HandlerImpl\Application\Application.Impl.cs:line 126
   at Microsoft.Maui.Controls.Application.Microsoft.Maui.IApplication.CreateWindow(IActivationState activationState) in D:\a\_work\1\s\src\Controls\src\Core\HandlerImpl\Application\Application.Impl.cs:line 40
   at Microsoft.Maui.Platform.ApplicationExtensions.CreatePlatformWindow(Activity activity, IApplication application, Bundle savedInstanceState) in D:\a\_work\1\s\src\Core\src\Platform\Android\ApplicationExtensions.cs:line 44
   at Microsoft.Maui.MauiAppCompatActivity.OnCreate(Bundle savedInstanceState) in D:\a\_work\1\s\src\Core\src\Platform\Android\MauiAppCompatActivity.cs:line 32
   at Android.App.Activity.n_OnCreate_Landroid_os_Bundle_(IntPtr jnienv, IntPtr native__this, IntPtr native_savedInstanceState) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net7.0/android-33/mcw/Android.App.Activity.cs:line 2816
   at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPL_V(_JniMarshal_PPL_V callback, IntPtr jnienv, IntPtr klazz, IntPtr p0) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 121
```

**Comment by rachelkang** [comment]
Hi, @vhugogarcia - are you able to reproduce this issue in a project that doesn't use Prism? If so, would you be able to share a repro?

**Comment by PureWeen** [comment]
@vhugogarcia your sample project gives me a 404

Most likely you have somewhere in your application where you are doing this
Application.Current.MainPage = new MainPage()

Or something of that nature. 

That type of initialization isn't compatible with Prism

**Comment by vhugogarcia** [comment]
Thanks @rachelkang for the faster response. Our friend Gerry created a few days ago a demo project on this repository: [https://github.com/gerry123/MauiMainPageSwitchingError.git](https://github.com/gerry123/MauiMainPageSwitchingError.git)

It is basically the same principle, if we change the MainPage (root) without a Navigation Page it breaks fully on Android.

Please find below a video about it. I'm open to jump into a screen share call to provide more details about it . I'm happy to collaborate with the team in anything I can!


https://user-images.githubusercontent.com/1047398/197225883-4967c37c-08a6-4048-ac3b-db923268343e.mp4

cc: @PureWeen 

Thank you

**Comment by vhugogarcia** [comment]
> PureWeen

I added permissions to the private repository to you Shane. Also added permissions to you @rachelkang 

Thanks for looking into.

**Comment by PureWeen** [comment]
@vhugogarcia there was one last big push in this space for [GA](https://github.com/dotnet/maui/pull/10530)

I was able to reproduce your issue on RC2.

Here it is running on the latest GA bits. 

I noticed in your video (after recording mine) that you clicked the icon. I also tested that path and same result as the video here. 

![2022-10-21_10-23-04](https://user-images.githubusercontent.com/5375137/197231816-41210145-59ea-47a6-8cfd-d2b148f647e9.gif)

**Comment by PureWeen** [comment]
@vhugogarcia you might hit that same exception when using a flyoutpage and swapping it in and out as well which should also be fixed by GA.

**Comment by vhugogarcia** [comment]
That is great to hear @PureWeen ,

So, I just want to clarify, you are saying that on the GA version expected to be released with .NET 7 final release on November this year, the issue we are experience should be fixed, right? 

I guess that for instance Prism.Maui issue should be solved automatically as well _(in-theory)_.

**Comment by PureWeen** [comment]
@vhugogarcia I chatted with @dansiegel about this one a little bit and pushed up some suggested paths. The root error here is caused by how Prism is wanting to inject into the windows collection. I tried to find a way to make that work but unfortunately, I think for now Prism will need to override `CreateWindow` to provide a `PrismWindow` and then once we implement the [factory](https://github.com/dotnet/maui/issues/8823).

**Comment by vhugogarcia** [comment]
Well, I'm basically swapping the MainPage because of this flow:

OnBoarding Screen (with a login button) -> Login button loads a modal to authenticate the user (Identity Server) -> Once authenticated the authenticator modal page is closed automatically (callback) and then on the OnBoarding screen I process the access token to retrieve the user information and then I navigate to the Dashboard page (Tabbed Page). 

However, when the user is at the Dashboard page, I should clear the navigation stack and Dashboard screen becomes the MainPage since I don't want them to navigate back to the login screen if they press a back button in Android or similar.

I will give it a try to Shell with RC2 and see how it works. 

Thanks @PureWeen for the insights! :)

**Comment by PureWeen** [comment]
@vhugogarcia yea, so you can just hide or show Shell parts to do that or just create different areas

```XAML
<Shell>
<ShellContent Route="LoginFlow" IsVisible={Binding IsAnonymous} />
<FlyoutItem IsVisible={Binding IsLoggedIn} ></Flyout>
```

Or just

```XAML
<Shell>
<ShellContent Route="LoginFlow" IsVisible={Binding IsAnonymous} />
<FlyoutItem/TabBar IsVisible={Binding IsLoggedIn} Route="LoggedIn":></Flyout>
```

Then when they log in
```C#
Shell.Current.GotoAsync("//LoggedIn")
```

Some variation of above

**Comment by gerry123** [comment]
I can confirm that #10271 is fixed in VS 17.4.1 (.net 7.0.100)...
